executor: show 'need reorg' and 'validating' in admin show ddl jobs's comment field

wjhuang2016 opened the pull request:
### What problem does this PR solve?

Issue Number: close #65196

Problem Summary:
The `COMMENT` field in `ADMIN SHOW DDL JOBS` output lacked clarity regarding whether a job was performing data reorganization or just validating existing data (common in lossy DDL optimizations).

### What changed and how does it work?

1.  Added `need reorg` label to the `COMMENT` field for jobs that involve data reorganization.
2.  Added `validating` label for jobs that perform data validation without reorganization (e.g., lossy DDL optimizations, varchar-to-char conversion).
3.  Introduced `getReorgAndVerifyLabels` helper to unify label logic for both main jobs and subjobs.
4.  Improved `MultiSchemaChange` summary logic to correctly aggregate `need reorg` and `validating` status from all subjobs.
5.  Ensured accurate reorg status for historical jobs by re-parsing `ModifyColumnArgs`.

### Check List

Tests <!-- At least one of them must be included. -->

- [X] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation
- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note
```release-note
Show "need reorg" and "validating" labels in ADMIN SHOW DDL JOBS's comment field for better observability.
```

wjhuang2016 committed:
executor: show 'need reorg' and 'validating' in admin show ddl jobs

wjhuang2016 committed:
executor: improve ddl job labels by using ReorgMeta as fallback

wjhuang2016 committed:
executor: show 'need reorg' and 'validating' using reorg meta

wjhuang2016 committed:
executor: show 'need reorg' and 'validating' using explicit persisted fields

wjhuang2016 committed:
executor: fix nogo errors and update unit tests

wjhuang2016 committed:
tests: add integration test for ddl job labels with masked dynamic columns

wjhuang2016 committed:
executor: restrict ddl job labels to classic kernel to avoid breaking next-gen tests

wjhuang2016 committed:
executor: restore ddl job labels tests and remove kernel restriction

wjhuang2016 committed:
executor: fix gofmt in show_ddl_jobs.go

wjhuang2016 committed:
tests: remove unstable add index test case from integration test

wjhuang2016 committed:
executor: refine admin show ddl jobs comments for next-gen compatibility

wjhuang2016 committed:
executor: handle validation fallback to reorg and add tests

wjhuang2016 committed:
pkg/executor: show analyze status in nextgen mode for add index

wjhuang2016 committed:
pkg/executor: skip 'need reorg' for add index in nextgen mode

wjhuang2016 committed:
pkg/executor: add comments for showCommentsFromJob and showCommentsFromSubjob

wjhuang2016 committed:
pkg/executor: always skip 'need reorg' for add index and primary key

joechenrh commented on pkg/ddl/modify_column.go:
```diff
@@ -324,8 +324,14 @@ func (w *worker) onModifyColumn(jobCtx *jobContext, job *model.Job) (ver int64,
 	switch args.ModifyColumnType {
 	case model.ModifyTypePrecheck:
+		if job.ReorgMeta != nil {
+			job.ReorgMeta.IsValidating = true
```
I prefer reset this to false after the job is finished. Otherwise, the comment will always contain "validating" even if the job is done.

joechenrh commented on pkg/executor/show_ddl_jobs.go:
```diff
@@ -299,12 +299,23 @@ func (e *DDLJobRetriever) appendJobToChunk(req *chunk.Chunk, job *model.Job, che
+	isAddingIndex := job.Type == model.ActionAddIndex ||
+		job.Type == model.ActionAddPrimaryKey
+	// For adding index, we don't need to show 'need reorg' to users.
+	if job.MayNeedReorg() && !isAddingIndex {
```
Maybe don't show "need reorg" for multi schema change too

joechenrh commented on pkg/meta/model/reorg.go:
```diff
@@ -101,6 +101,8 @@ type DDLReorgMeta struct {
 	Concurrency   atomic.Int64 `json:"concurrency"`
 	BatchSize     atomic.Int64 `json:"batch_size"`
 	MaxWriteSpeed atomic.Int64 `json:"max_write_speed"`
+	// IsValidating indicates whether this job is validating the data.
```
Store it in `SubJob`, so we can show validation status in each sub job.

wjhuang2016 committed:
executor: address review comments for PR 65197

wjhuang2016 committed:
executor: only show DXF labels when using distributed reorg

wjhuang2016 committed:
executor: don't show DXF labels for modify column jobs

wjhuang2016 committed:
Merge branch 'master' of https://github.com/pingcap/tidb into admin-show-ddl-jobs-labels

wjhuang2016 committed:
executor: fix panic in MayNeedReorg and ensure DXF labels logic

wjhuang2016 committed:
executor: address review comments for PR 65197

- Restore simple proxy job creation in MayNeedReorg for ActionMultiSchemaChange
- Use Job{Type: sub.Type, NeedReorg: sub.NeedReorg} instead of sub.ToProxyJob(job, i)
- Reason: At job submission time, modify type is not determined yet, so we need to use NeedReorg as a rough judgment

wjhuang2016 commented:
/retest
